src/persist-client/src/operators/shard_source.rs

```diff
@@ -317,209 +302,116 @@ where
             }
         };
 
-        let create_subscribe = async {
-            read.subscribe(as_of.clone()).await
-        };
-        tokio::pin!(create_subscribe);
-        let token_is_dropped = token_tx.closed();
-        tokio::pin!(token_is_dropped);
-
-        // creating a Subscribe can take indefinitely long (e.g. as_of in the far future),
-        // so we race its creation with our token to ensure any async work is dropped once
-        // the token is dropped.
-        //
-        // NB: reading from a channel (token_is_dropped) is cancel-safe.
-        //     create_subscribe is NOT cancel-safe, but we will not retry it if it is dropped.
-        let subscription = match futures::future::select(token_is_dropped, create_subscribe).await {
-            Either::Left(_) => {
-                // the token dropped before we finished creating our Subscribe.
-                // we can return immediately, as we could not have emitted any
-                // parts to fetch.
-                cap_set.downgrade(&[]);
-                return;
-            }
-            Either::Right((subscription, _)) => {
-                subscription
-            }
-        };
-
-        let mut subscription = subscription.unwrap_or_else(|e| {
+        let subscription = read.subscribe(as_of.clone()).await.unwrap_or_else(|e| {
             panic!(
                 "{}: {} cannot serve requested as_of {:?}: {:?}",
                 name_owned, shard_id, as_of, e
             )
         });
 
-        let mut batch_parts = vec![];
+        // We're about to start producing parts to be fetched whose leases will be returned by the
+        // `shard_source_descs_return` operator above. In order for that operator to successfully
+        // return the leases we send it two things:
+        // 1. A type erased Rc<RefCell<Subscribe>>
+        // 2. The associated SubscriptionLeaseReturner
+        // This ensures that that `shard_source_descs_return` can keep our Subscribe handle alive
+        // after this operator has shutdown, which in turns prevents its SeqNo hold from expiring.
+        // The reason for the type erasure is to prevent that operator from attempting a borrow_mut
+        // etc. We just need it to hold it until it has finished returning all the outstanding parts.
+        let lease_returner = subscription.lease_returner().clone();
+        let subscription = Rc::new(RefCell::new(subscription));
+        #[allow(clippy::as_conversions)]
+        let sub_handle = Rc::clone(&subscription) as Rc<dyn Any>;
```

Comment:
I don't have a good sense of how sketchy we think this cast is, but it seems like we could avoid it for ~almost free by `Box`ing the handle.

Reply:
the cast is fine, its the same thing as the coercion that happens with boxing it